### DISCUSSION TITLE
Specular Aliasing

### DISCUSSION THREAD

**Discussion by harrycollin** [pmndrs/postprocessing#650] [Q&A]
I'm encountering some issues with what I believe is called 'Specular Aliasing'. After many failed attempts to tweak SMAA & MSAA parameters I'm a bit stuck on how to solve this. The image attached showcases the issue quite well, the white edges on the front of the car appear even after using MSAA 8X with Ultra settings. 
![image](https://github.com/user-attachments/assets/df1250e6-b79f-4c8f-b3f0-f9adcb78ffed)

A similar problem has been talked about over [here](https://discourse.threejs.org/t/antialiasing-problem-thin-light-reflections-on-objects-edges-look-jagged/30713/11) but doesn't seem to have a concrete solution. I'm open to any suggestions :)

**Comment by vanruesc** [comment]
This doesn't look like it's related to `postprocessing`. Does the problem also occur without using `postprocessing`?

**Comment by harrycollin** [comment]
Apologies for the lack of information & structure in my original issue. It was intended to be more of an exploratory issue/discussion rather than something directly inheritably wrong with `postprocessing`. Perhaps I should have posted over here. 

The aliasing does indeed appear without using `postprocessing`. I'm not sure what causes this to happen, only that it's more obvious on ultra smooth surfaces in scenes that use HDRs with high contrast. 

I would have gladly filled out the original issue with more concrete cause/effect information, hoverer I'm not that knowledgeable on this subject.

**Reply by vanruesc** [reply]
No problem :+1: 

The aliasing artifacts you're seeing are likely caused by the lighting setup in your scene, including the HDR background which is probably being used as a `scene.environment`.

I think the problem is that MSAA, and any other antialiasing algorithm, doesn't work with very bright pixels.

Lights with high intensities cause pixel values to exceed the LDR range [0, 1]. Antialiasing is a convolution/blurring technique that combines neighboring pixels into one to create smooth linear gradients. This works well if the pixels are in the LDR range because then the resulting gradients can be displayed on screen. However, if any one of the pixels involved in the blurring exceeds the value 1, the result will be more likely to lie outside of the renderable LDR range. And since you have a lot of very bright highlights along geometric edges in your scene, almost all of those pixels probably exceed the LDR range.

You can tune the lighting and the material settings of your objects to combat this. The issue should also go away if you use three's inline tone mapping (e.g. `renderer.toneMapping = AgXToneMapping`), but this approach is incompatible with postprocessing in general.

**Reply by vanruesc** [reply]
See also https://mynameismjp.wordpress.com/2012/10/24/msaa-overview/ for more infos on MSAA. The section "Working with HDR and Tone Mapping" covers this topic as well.

**Reply by harrycollin** [reply]
Thank you for the comprehensive breakdown. I've read through and have a better understand of the problem. The inline ```toneMapping``` does indeed work, but isn't good for postprocessing, I'm guessing this is because the pixels fed to postprocessing is already in LDR by that point? I was under the impression I could use the ```ToneMappingEffect``` before using ```SMAAEffect```, but it seems my understand of how the effects pass data to each other is wrong. I wrongly assumed that the pixel data after ```ToneMappingEffect``` would be converted to LDR.

**Reply by vanruesc** [reply]
> I'm guessing this is because the pixels fed to postprocessing is already in LDR by that point?

Correct.

> I was under the impression I could use the ToneMappingEffect before using SMAAEffect, but it seems my understand of how the effects pass data to each other is wrong.

That's a fair assumption and it's actually how it should be done. However, while it's technically possible to merge AA with tone mapping in one pass, it's not guaranteed that all jagged lines will be properly smoothed this way. A better solution is to have one `EffectPass` that performs tone mapping and a separate final `EffectPass` that applies antialiasing based on the LDR output of the previous pass.

Still, this won't solve the original issue of broken lines because doing AA in post is image-based while MSAA is geometry based. MSAA actually renders and blends _additional_ pixels around geometric edges and that's something postprocesing can't do. SSAA is the closest thing to MSAA, but it's prohibitively expensive. Other AA such as FXAA and SMAA won't be able to reconstruct the broken lines.

>  I wrongly assumed that the pixel data after ToneMappingEffect would be converted to LDR.

When tone mapping and AA are merged into one pass, the AA algorithm will still operate on the input image of the pass which won't be affected by the tone mapping. The AA effect will pull in pixels that are not tone mapped and it will blend them with the pixel that is currently going through the combined effect shader.

**Comment by Arecsu** [comment]
Hey! Have you found a workaround for these particular form of specular aliasing? I think the same as @vanruesc in which the problem seems to be very bright values outside of the LDR range. But so far I couldn't find a proper way to mitigate this issue it in any way :/

**Reply by vanruesc** [reply]
I looked around for solutions a while ago and found this https://www.shadertoy.com/view/WssyR7 which may imply that this type of aliasing can be fixed on a lower level, i.e. in the shaders used to render the meshes. I haven't looked deeper into this though, because this type of antialiasing doesn't really address the issue of very bright pixels. But maybe something can be done in three's PBR shaders to fix some of the artifacts.

**Reply by Arecsu** [reply]
Thank you so much for your answer! In fact, I tried to implement all sorts of Specular Antialiasing techniques by patching the PBR shader three.js uses. Tried everything: Square Enix papers and implementations, Valve's VR Specular Antialiasing, including your shadertoy example. Although it does have some effect in curved geometry, I could never make it work in 90 degree edges in my object, even with some bevel applied to them.

I'm not that knowledgeable when it comes to shaders by the way, so that's a bummer. However, even if there was a way to implement it, I do think the elephant in the room is indeed the huge contrast differences in the HDR range of radiance returned in some angles from the object. The aliased lines could be very, very bright and the rest of the model have black values which makes any antialiasing algorithm to not understand what to smooth. Maybe there's a way to reduce that radiance at certain viewing angles as a way to sort of mitigate the issue and let MSAA do its job. But I don't have any idea how to do it honestly. 

Also tried to see if there's a positive effect in feeding the postprocessing pipeline with an already tonemapped buffer. It does help, just a bit! So I think your suspicious are correct regarding HDR ranges. But is not only insufficient if the contrast in the specular is still high, it also breaks the pipeline in a way that other effects won't have the same quality as feeding a HDR image from the beginning.

https://arecsu.github.io/monk-viewer/ → here is a working example using the alpha v7 postprocessing library with a full HDR pipeline where the edges of the artwork presents this issue.

For my part, I have already decided to call for surrender and wait for a good TAA implementation in hopes it will get around these damn white lines. MSAA at this point, even in modern graphics, is not enough to tackle these issues. Not even Normal/Displacement Map aliasing issues that could exist at some point.

By the way, love this postprocessing library. I'm having such a great time using it, particularly the v7 alpha :). My previous linked demo uses offscreen canvas rendering and so far, nothing broke. Might be placebo but I think it works better than v6! Do you need help to get some effects/shaders ported to it?

**Reply by vanruesc** [reply]
> wait for a good TAA implementation

Yeah, TAA would fix this type of aliasing, but it comes with its own issues like ghosting and overblurring. When I get to it, I'll try to implement a version that works with minimal frame accumulation.

> Might be placebo but I think it works better than v6!

Well, v7 is supposed to work better than v6 after all. It would be bad if it didn't :sweat_smile: 

> Do you need help to get some effects/shaders ported to it?

There's a list of open tasks in #600.
